File: src/compare.ts
import type {
  AllowedDelta,
  AllowedDifference,
  CompareField,
  ComparisonResult,
  Finding,
  JsonValueType,
  ProbeReceipt,
  RunReceipt,
  TheseusContract,
} from "./types.js";

export function compareReceipts(contract: TheseusContract, receipts: RunReceipt[]): ComparisonResult {
  const findings: Finding[] = [];
  const allowedDifferences: AllowedDifference[] = [];

  for (const receipt of receipts) {
    for (const probe of contract.probes) {
      const observed = receipt.probes.find((candidate) => candidate.id === probe.id);
      if (!observed) continue;
      if (probe.expect?.exit !== undefined && observed.exitCode !== probe.expect.exit) {
        findings.push({
          code: "THS-EXPECT-001",
          probe: probe.id,
          field: "exit",
          profiles: [receipt.profile],
          message: `${receipt.profile}/${probe.id}: expected exit ${probe.expect.exit}, observed ${String(observed.exitCode)}`,
          expected: probe.expect.exit,
          actual: observed.exitCode,
        });
      }
      if (probe.expect?.stdoutJson) {
        checkJsonExpectations(probe.id, receipt.profile, observed, probe.expect.stdoutJson, findings);
      }
    }
  }

  for (let leftIndex = 0; leftIndex < receipts.length; leftIndex += 1) {
    for (let rightIndex = leftIndex + 1; rightIndex < receipts.length; rightIndex += 1) {
      const left = receipts[leftIndex]!;
      const right = receipts[rightIndex]!;
      for (const probe of contract.probes) {
        const leftProbe = left.probes.find((candidate) => candidate.id === probe.id);
        const rightProbe = right.probes.find((candidate) => candidate.id === probe.id);
        if (!leftProbe || !rightProbe) continue;
        for (const field of probe.compare) {
          const values = fieldValues(field, left, right, leftProbe, rightProbe);
          if (Object.is(values.left, values.right)) continue;
          const allowed = contract.allowedDeltas.find((delta) => matchesAllowed(delta, probe.id, field, left.profile, right.profile, values.left, values.right));
          if (allowed) {
            allowedDifferences.push({ probe: probe.id, field, profiles: [left.profile, right.profile], reason: allowed.reason });
            continue;
          }
          findings.push(findingFor(field, probe.id, left.profile, right.profile, values));
        }
      }
    }
  }

  return { findings, allowedDifferences };
}

function fieldValues(
  field: CompareField,
  left: RunReceipt,
  right: RunReceipt,
  leftProbe: ProbeReceipt,
  rightProbe: ProbeReceipt,
): { left: number | string | null; right: number | string | null; displayLeft: unknown; displayRight: unknown } {
  if (field === "exit") {
    return { left: leftProbe.exitCode, right: rightProbe.exitCode, displayLeft: leftProbe.exitCode, displayRight: rightProbe.exitCode };
  }
  if (field === "stdout") {
    return { left: leftProbe.stdoutSha256, right: rightProbe.stdoutSha256, displayLeft: leftProbe.stdout, displayRight: rightProbe.stdout };
  }
  if (field === "stderr") {
    return { left: leftProbe.stderrSha256, right: rightProbe.stderrSha256, displayLeft: leftProbe.stderr, displayRight: rightProbe.stderr };
  }
  return { left: left.runtime.sha256, right: right.runtime.sha256, displayLeft: left.runtime.canonical, displayRight: right.runtime.canonical };
}

function matchesAllowed(
  delta: AllowedDelta,
  probe: string,
  field: CompareField,
  leftProfile: string,
  rightProfile: string,
  left: number | string | null,
  right: number | string | null,
): boolean {
  if (delta.probe !== probe || delta.field !== field) return false;
  if (delta.between[0] === leftProfile && delta.between[1] === rightProfile) {
    return field === "exit" ? Object.is(delta.from, left) && Object.is(delta.to, right) : delta.fromSha256 === left && delta.toSha256 === right;
  }
  if (delta.between[0] === rightProfile && delta.between[1] === leftProfile) {
    return field === "exit" ? Object.is(delta.from, right) && Object.is(delta.to, left) : delta.fromSha256 === right && delta.toSha256 === left;
  }
  return false;
}

function findingFor(
  field: CompareField,
  probe: string,
  leftProfile: string,
  rightProfile: string,
  values: { left: number | string | null; right: number | string | null; displayLeft: unknown; displayRight: unknown },
): Finding {
  const profiles = [leftProfile, rightProfile];
  if (field === "exit") {
    return {
      code: "THS-PARITY-001",
      probe,
      field,
      profiles,
      message: `${probe}: exit differs between ${leftProfile} and ${rightProfile}`,
      actual: { [leftProfile]: values.displayLeft, [rightProfile]: values.displayRight },
    };
  }
  const code = field === "runtime" ? "THS-RUNTIME-001" : "THS-PARITY-002";
  return {
    code,
    probe,
    field,
    profiles,
    message: `${probe}: ${field} differs between ${leftProfile} and ${rightProfile}`,
    actual: { [leftProfile]: values.displayLeft, [rightProfile]: values.displayRight },
    digests: { [leftProfile]: String(values.left), [rightProfile]: String(values.right) },
  };
}

function checkJsonExpectations(
  probe: string,
  profile: string,
  receipt: ProbeReceipt,
  expectations: Array<{ path: string; type: JsonValueType }>,
  findings: Finding[],
): void {
  let parsed: unknown;
  try {
    parsed = JSON.parse(receipt.stdout);
  } catch {
    findings.push({
      code: "THS-EXPECT-002",
      probe,
      field: "stdoutJson",
      profiles: [profile],
      message: `${profile}/${probe}: stdout is not valid JSON`,
    });
    return;
  }
  for (const expectation of expectations) {
    const value = getPath(parsed, expectation.path);
    const actualType = jsonType(value);
    if (actualType !== expectation.type) {
      findings.push({
        code: "THS-EXPECT-002",
        probe,
        field: "stdoutJson",
        profiles: [profile],
        message: `${profile}/${probe}: JSON path ${expectation.path} expected ${expectation.type}, observed ${actualType}`,
        expected: expectation.type,
        actual: actualType,
      });
    }
  }
}

function getPath(value: unknown, path: string): unknown {
  const segments = path.startsWith("/") ? parseJsonPointer(path) : path.split(".");
  if (!segments) return undefined;
  let current = value;
  for (const segment of segments) {
    if (current === null || typeof current !== "object" || !Object.hasOwn(current, segment)) return undefined;
    current = (current as Record<string, unknown>)[segment];
  }
  return current;
}

function parseJsonPointer(path: string): string[] | null {
  const segments: string[] = [];
  for (const token of path.slice(1).split("/")) {
    if (/~(?:[^01]|$)/.test(token)) return null;
    segments.push(token.replaceAll("~1", "/").replaceAll("~0", "~"));
  }
  return segments;
}

function jsonType(value: unknown): JsonValueType | "undefined" {
  if (value === undefined) return "undefined";
  if (value === null) return "null";
  if (Array.isArray(value)) return "array";
  return typeof value as JsonValueType;
}

File: internal/compare/compare.go
package compare

import (
	"bytes"
	"encoding/json"
	"fmt"
	"reflect"
	"strings"

	"github.com/JeremyDev87/theseus/internal/model"
)

func Receipts(contract model.Contract, receipts []model.RunReceipt) model.ComparisonResult {
	result := model.ComparisonResult{Findings: []model.Finding{}, AllowedDifferences: []model.AllowedDifference{}}
	for _, receipt := range receipts {
		for _, probe := range contract.Probes {
			observed := findProbe(receipt.Probes, probe.ID)
			if observed == nil {
				continue
			}
			if probe.Expect != nil && probe.Expect.Exit != nil && !equalIntPointer(observed.ExitCode, probe.Expect.Exit) {
				result.Findings = append(result.Findings, model.Finding{
					Code: "THS-EXPECT-001", Probe: probe.ID, Field: "exit", Profiles: []string{receipt.Profile},
					Message:  fmt.Sprintf("%s/%s: expected exit %d, observed %s", receipt.Profile, probe.ID, *probe.Expect.Exit, displayExit(observed.ExitCode)),
					Expected: *probe.Expect.Exit, Actual: exitValue(observed.ExitCode),
				})
			}
			if probe.Expect != nil && len(probe.Expect.StdoutJSON) > 0 {
				checkJSON(probe.ID, receipt.Profile, *observed, probe.Expect.StdoutJSON, &result.Findings)
			}
		}
	}
	for leftIndex := 0; leftIndex < len(receipts); leftIndex++ {
		for rightIndex := leftIndex + 1; rightIndex < len(receipts); rightIndex++ {
			left, right := receipts[leftIndex], receipts[rightIndex]
			for _, probe := range contract.Probes {
				leftProbe, rightProbe := findProbe(left.Probes, probe.ID), findProbe(right.Probes, probe.ID)
				if leftProbe == nil || rightProbe == nil {
					continue
				}
				for _, field := range probe.Compare {
					leftValue, rightValue, leftDisplay, rightDisplay := fieldValues(field, left, right, *leftProbe, *rightProbe)
					if reflect.DeepEqual(leftValue, rightValue) {
						continue
					}
					if allowed := findAllowed(contract.AllowedDeltas, probe.ID, field, left.Profile, right.Profile, leftValue, rightValue); allowed != nil {
						result.AllowedDifferences = append(result.AllowedDifferences, model.AllowedDifference{Probe: probe.ID, Field: field, Profiles: [2]string{left.Profile, right.Profile}, Reason: allowed.Reason})
						continue
					}
					result.Findings = append(result.Findings, finding(field, probe.ID, left.Profile, right.Profile, leftValue, rightValue, leftDisplay, rightDisplay))
				}
			}
		}
	}
	return result
}

func findProbe(probes []model.ProbeReceipt, id string) *model.ProbeReceipt {
	for index := range probes {
		if probes[index].ID == id {
			return &probes[index]
		}
	}
	return nil
}

func fieldValues(field model.CompareField, left, right model.RunReceipt, leftProbe, rightProbe model.ProbeReceipt) (any, any, any, any) {
	switch field {
	case model.FieldExit:
		return exitValue(leftProbe.ExitCode), exitValue(rightProbe.ExitCode), exitValue(leftProbe.ExitCode), exitValue(rightProbe.ExitCode)
	case model.FieldStdout:
		return leftProbe.StdoutSHA256, rightProbe.StdoutSHA256, leftProbe.Stdout, rightProbe.Stdout
	case model.FieldStderr:
		return leftProbe.StderrSHA256, rightProbe.StderrSHA256, leftProbe.Stderr, rightProbe.Stderr
	default:
		return left.Runtime.SHA256, right.Runtime.SHA256, left.Runtime.Canonical, right.Runtime.Canonical
	}
}

func findAllowed(deltas []model.AllowedDelta, probe string, field model.CompareField, leftProfile, rightProfile string, left, right any) *model.AllowedDelta {
	for index := range deltas {
		delta := &deltas[index]
		if delta.Probe != probe || delta.Field != field {
			continue
		}
		if delta.Between[0] == leftProfile && delta.Between[1] == rightProfile && allowedValues(*delta, left, right) {
			return delta
		}
		if delta.Between[0] == rightProfile && delta.Between[1] == leftProfile && allowedValues(*delta, right, left) {
			return delta
		}
	}
	return nil
}

func allowedValues(delta model.AllowedDelta, from, to any) bool {
	if delta.Field == model.FieldExit {
		return reflect.DeepEqual(nullableValue(delta.From), from) && reflect.DeepEqual(nullableValue(delta.To), to)
	}
	return delta.FromSHA256 == fmt.Sprint(from) && delta.ToSHA256 == fmt.Sprint(to)
}

func nullableValue(value *int) any {
	if value == nil {
		return nil
	}
	return *value
}

func finding(field model.CompareField, probe, leftProfile, rightProfile string, left, right, displayLeft, displayRight any) model.Finding {
	profiles := []string{leftProfile, rightProfile}
	actual := map[string]any{leftProfile: displayLeft, rightProfile: displayRight}
	if field == model.FieldExit {
		return model.Finding{Code: "THS-PARITY-001", Probe: probe, Field: string(field), Profiles: profiles, Message: fmt.Sprintf("%s: exit differs between %s and %s", probe, leftProfile, rightProfile), Actual: actual}
	}
	code := "THS-PARITY-002"
	if field == model.FieldRuntime {
		code = "THS-RUNTIME-001"
	}
	return model.Finding{Code: code, Probe: probe, Field: string(field), Profiles: profiles, Message: fmt.Sprintf("%s: %s differs between %s and %s", probe, field, leftProfile, rightProfile), Actual: actual, Digests: map[string]string{leftProfile: fmt.Sprint(left), rightProfile: fmt.Sprint(right)}}
}

func checkJSON(probe, profile string, receipt model.ProbeReceipt, expectations []model.JSONPathExpectation, findings *[]model.Finding) {
	var parsed any
	if !json.Valid([]byte(receipt.Stdout)) {
		*findings = append(*findings, model.Finding{Code: "THS-EXPECT-002", Probe: probe, Field: "stdoutJson", Profiles: []string{profile}, Message: fmt.Sprintf("%s/%s: stdout is not valid JSON", profile, probe)})
		return
	}
	decoder := json.NewDecoder(bytes.NewBufferString(receipt.Stdout))
	decoder.UseNumber()
	if err := decoder.Decode(&parsed); err != nil {
		*findings = append(*findings, model.Finding{Code: "THS-EXPECT-002", Probe: probe, Field: "stdoutJson", Profiles: []string{profile}, Message: fmt.Sprintf("%s/%s: stdout is not valid JSON", profile, probe)})
		return
	}
	for _, expectation := range expectations {
		value, found := getPath(parsed, expectation.Path)
		actualType := jsonType(value, found)
		if actualType != string(expectation.Type) {
			*findings = append(*findings, model.Finding{Code: "THS-EXPECT-002", Probe: probe, Field: "stdoutJson", Profiles: []string{profile}, Message: fmt.Sprintf("%s/%s: JSON path %s expected %s, observed %s", profile, probe, expectation.Path, expectation.Type, actualType), Expected: string(expectation.Type), Actual: actualType})
		}
	}
}

func getPath(value any, path string) (any, bool) {
	segments := strings.Split(path, ".")
	if strings.HasPrefix(path, "/") {
		var ok bool
		segments, ok = jsonPointer(path)
		if !ok {
			return nil, false
		}
	}
	current := value
	for _, segment := range segments {
		switch typed := current.(type) {
		case map[string]any:
			var ok bool
			current, ok = typed[segment]
			if !ok {
				return nil, false
			}
		case []any:
			if segment == "length" {
				current = float64(len(typed))
				continue
			}
			index, ok := arrayIndex(segment, len(typed))
			if !ok {
				return nil, false
			}
			current = typed[index]
		default:
			return nil, false
		}
	}
	return current, true
}

func jsonPointer(path string) ([]string, bool) {
	tokens := strings.Split(path[1:], "/")
	for index, token := range tokens {
		var builder strings.Builder
		for offset := 0; offset < len(token); offset++ {
			if token[offset] != '~' {
				builder.WriteByte(token[offset])
				continue
			}
			if offset+1 >= len(token) || (token[offset+1] != '0' && token[offset+1] != '1') {
				return nil, false
			}
			offset++
			if token[offset] == '0' {
				builder.WriteByte('~')
			} else {
				builder.WriteByte('/')
			}
		}
		tokens[index] = builder.String()
	}
	return tokens, true
}

func arrayIndex(value string, length int) (int, bool) {
	if value == "" || (len(value) > 1 && value[0] == '0') {
		return 0, false
	}
	index := 0
	for _, char := range value {
		if char < '0' || char > '9' {
			return 0, false
		}
		index = index*10 + int(char-'0')
	}
	return index, index >= 0 && index < length
}

func jsonType(value any, found bool) string {
	if !found {
		return "undefined"
	}
	if value == nil {
		return "null"
	}
	switch value.(type) {
	case string:
		return "string"
	case float64, json.Number:
		return "number"
	case bool:
		return "boolean"
	case []any:
		return "array"
	case map[string]any:
		return "object"
	default:
		return "undefined"
	}
}

func equalIntPointer(left, right *int) bool {
	return reflect.DeepEqual(exitValue(left), exitValue(right))
}
func exitValue(value *int) any {
	if value == nil {
		return nil
	}
	return *value
}
func displayExit(value *int) string {
	if value == nil {
		return "null"
	}
	return fmt.Sprint(*value)
}
